File: app/src/main/java/com/tuvakov/zeyoube/android/VideoFeedAdapter.java
package com.tuvakov.zeyoube.android;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.tuvakov.zeyoube.android.data.Video;

import org.threeten.bp.format.DateTimeFormatter;
import org.threeten.bp.format.FormatStyle;


public class VideoFeedAdapter extends ListAdapter<Video, VideoFeedAdapter.VideoViewHolder> {

    private static final String TAG = "VideoFeedAdapter";
    private ItemClickListener mItemClickListener;

    public VideoFeedAdapter() {
        super(new DiffUtil.ItemCallback<Video>() {
            @Override
            public boolean areItemsTheSame(@NonNull Video oldItem, @NonNull Video newItem) {
                return oldItem.getId() == newItem.getId();
            }

            @Override
            public boolean areContentsTheSame(@NonNull Video oldItem, @NonNull Video newItem) {
                return oldItem.getTitle().equals(newItem.getTitle())
                        && oldItem.getDescription().equals(newItem.getDescription())
                        && oldItem.getVideoId().equals(newItem.getVideoId());
            }
        });
    }

    @NonNull
    @Override
    public VideoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_video_feed, parent, false);
        return new VideoViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull VideoViewHolder holder, int position) {
        Video currentVideo = getItem(position);

        Glide.with(holder.imageViewChannelAvatar.getContext())
                .load(currentVideo.getChannelAvatar())
                .centerCrop()
                .circleCrop()
                .placeholder(new ColorDrawable(Color.GRAY))
                .into(holder.imageViewChannelAvatar);

        Glide.with(holder.imageViewVideoThumbnail.getContext())
                .load(currentVideo.getThumbnail())
                .centerCrop()
                .placeholder(new ColorDrawable(Color.GRAY))
                .into(holder.imageViewVideoThumbnail);

        holder.textViewInfo.setText(String.format("%s \u2022 ", currentVideo.getChannelTitle()));
        holder.textViewTitle.setText(currentVideo.getTitle());
        String publishedAt = currentVideo.getLocalDateTimePublishedAt()
                .format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM));
        Log.d(TAG, "onBindViewHolder: " + currentVideo.getLocalDateTimePublishedAt());
        holder.textViewInfo.append(publishedAt);
    }


    public Video getVideoAt(int i) {
        return getItem(i);
    }


    public void setItemClickListener(ItemClickListener itemClickListener) {
        this.mItemClickListener = itemClickListener;
    }

    class VideoViewHolder extends RecyclerView.ViewHolder {

        private ImageView imageViewVideoThumbnail;
        private TextView textViewTitle;
        private TextView textViewInfo;
        private ImageView imageViewChannelAvatar;

        VideoViewHolder(@NonNull View itemView) {
            super(itemView);
            imageViewVideoThumbnail = itemView.findViewById(R.id.iv_video_thumbnail);
            textViewTitle = itemView.findViewById(R.id.tv_video_title);
            textViewInfo = itemView.findViewById(R.id.tv_video_info);
            imageViewChannelAvatar = itemView.findViewById(R.id.iv_channel_avatar);

            itemView.setOnClickListener(view -> {
                int position = getAdapterPosition();
                if (mItemClickListener != null && position != RecyclerView.NO_POSITION) {
                    mItemClickListener.onClick(getItem(position));
                }
            });
        }
    }

    public interface ItemClickListener {
        void onClick(Video note);
    }
}

File: app/src/main/java/com/tuvakov/zetube/android/ui/videos/VideoFeedAdapter.kt
package com.tuvakov.zetube.android.ui.videos

import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.tuvakov.zetube.android.R
import com.tuvakov.zetube.android.data.Video
import com.tuvakov.zetube.android.databinding.ItemVideoFeedBinding
import com.tuvakov.zetube.android.ui.channel.ChannelDetailActivity
import com.tuvakov.zetube.android.ui.player.PlayerActivity

class VideoFeedAdapter : ListAdapter<Video, VideoViewHolder>(VIDEO_COMPARATOR) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VideoViewHolder {
        val binding = ItemVideoFeedBinding.inflate(LayoutInflater.from(parent.context))
        return VideoViewHolder(binding)
    }

    override fun onBindViewHolder(holder: VideoViewHolder, position: Int) {
        val currentVideo = getItem(position)
        if (currentVideo != null) {
            holder.bind(currentVideo)
        }
    }

    companion object {
        private const val TAG = "VideoFeedAdapter"
        private val VIDEO_COMPARATOR = object : DiffUtil.ItemCallback<Video>() {
            override fun areItemsTheSame(oldItem: Video, newItem: Video): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: Video, newItem: Video): Boolean {
                return oldItem.title == newItem.title && oldItem.description == newItem.description
                        && oldItem.publishedAt == newItem.publishedAt
            }
        }
    }
}

class VideoViewHolder(
        private val vBinding: ItemVideoFeedBinding
) : RecyclerView.ViewHolder(vBinding.root) {

    private lateinit var video: Video

    init {

        /* Open PlayerActivity on Video item click */
        itemView.setOnClickListener {
            val intent = Intent(it.context, PlayerActivity::class.java)
            intent.putExtra(PlayerActivity.KEY_EXTRA_VIDEO_ID, video.id)
            it.context.startActivity(intent)
        }

        /* Open ChannelDetailActivity on channel avatar click  */
        vBinding.ivChannelAvatar.setOnClickListener {
            // If already in ChannelDetailActivity then don't relaunch
            if (it.context !is VideoFeedActivity) {
                return@setOnClickListener
            }
            val intent = Intent(it.context, ChannelDetailActivity::class.java)
            intent.putExtra(ChannelDetailActivity.EXTRA_CHANNEL_ID, video.channelId)
            intent.putExtra(ChannelDetailActivity.EXTRA_CHANNEL_TITLE, video.channelTitle)
            it.context.startActivity(intent)
        }

        /* Show pop-up menu on 'more' icon click */
        vBinding.ivMore.setOnClickListener { view: View ->
            val popup = PopupMenu(view.context, view)
            popup.setOnMenuItemClickListener { item: MenuItem ->
                if (item.itemId == R.id.menu_item_share) {
                    val videoLink = PlayerActivity.VIDEO_BASE_LINK + video.id
                    val intent = Intent(Intent.ACTION_SEND)
                    intent.type = "text/plain"
                    intent.putExtra(Intent.EXTRA_TEXT, videoLink)
                    intent.resolveActivity(view.context.packageManager)?.let {
                        view.context.startActivity(intent)
                    }
                    return@setOnMenuItemClickListener true
                }
                false
            }
            popup.inflate(R.menu.popup_menu_feed_item)
            popup.show()
        }
    }

    fun bind(video: Video) {
        this.video = video
        Glide.with(vBinding.root.context)
                .load(video.channelAvatar)
                .centerCrop()
                .circleCrop()
                .placeholder(ColorDrawable(Color.GRAY))
                .into(vBinding.ivChannelAvatar)

        Glide.with(vBinding.root.context)
                .load(video.thumbnail)
                .centerCrop()
                .placeholder(ColorDrawable(Color.GRAY))
                .into(vBinding.ivVideoThumbnail)

        vBinding.tvVideoInfo.text = String.format("%s \u2022 ", video.channelTitle)
        vBinding.tvVideoTitle.text = video.title
        vBinding.tvVideoInfo.append(DateUtils.getRelativeTimeSpanString(video.publishedAt))
    }
}